src/core/lib/channel/channel_args_preconditioning.h

```diff
@@ -0,0 +1,62 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_CORE_LIB_CHANNEL_CHANNEL_ARGS_PRECONDITIONING_H
+#define GRPC_CORE_LIB_CHANNEL_CHANNEL_ARGS_PRECONDITIONING_H
+
+#include <grpc/support/port_platform.h>
+
+#include <functional>
+#include <vector>
+
+#include "src/core/lib/channel/channel_args.h"
+
+namespace grpc_core {
+
+// Registry of mutators for channel args.
+// Surface APIs should call into this with channel args received from outside
+// of gRPC, in order to prepare those channel args for the expections of the
+// gRPC internals.
+class ChannelArgsPreconditioning {
+ public:
+  // Take channel args and mutate them.
+  // Does not take ownership of the channel args passed in.
+  // Returns a new channel args object that is owned by the caller.
+  using Stage =
```

Comment:
is the multi-stage process gonna be bad for performance? (since we are gonna be copying the channel args at each stage). Can the answer to this question also be included as a comment here so that future readers know that it was considered?

Reply:
Going to answer inline since the answer changes soon: right now, it's not great, but equivalent to what we have.
Once we switch to the C++ representation with AVL, we'll be able to pass the object directly through from stage to stage and not do complete copies.